Sources/PackageCollections/PackageIndex.swift

```diff
@@ -0,0 +1,186 @@
+
+    func listPackages(
+        offset: Int,
+        limit: Int,
+        callback: @escaping (Result<PackageCollectionsModel.PaginatedPackageList, Error>) -> Void
+    ) {
+        self.runIfConfigured(callback: callback) { _, _ in
+            // TODO: cap `limit`
+            // TODO: call package index's list API
+            fatalError("Not implemented: \(#function)")
+        }
+    }
+
+    private func runIfConfigured<T>(
+        callback: @escaping (Result<T, Error>) -> Void,
+        closure: @escaping (Foundation.URL, (Result<T, Error>) -> Void) -> Void
+    ) {
+        let callback = self.makeAsync(callback)
+        
+        guard self.enabled else {
+            return callback(.failure(PackageIndexError.featureDisabled))
+        }
+        
+        self.get { result in
+            switch result {
+            case .failure(let error):
+                callback(.failure(error))
+            case .success(.none):
+                callback(.failure(PackageIndexError.notConfigured))
+            case .success(.some(let url)):
+                closure(url, callback)
+            }
+        }
+    }
+
+    private func makeAsync<T>(_ closure: @escaping (Result<T, Error>) -> Void) -> (Result<T, Error>) -> Void {
+        { result in self.callbackQueue.async { closure(result) } }
+    }
+}
+
+// MARK: - PackageMetadataProvider conformance
+
+extension PackageIndex: PackageMetadataProvider {
+    func get(
+        identity: PackageIdentity,
+        location: String,
+        callback: @escaping (Result<PackageCollectionsModel.PackageBasicMetadata, Error>, PackageMetadataProviderContext?) -> Void
+    ) {
+        self.getPackageMetadata(identity: identity, location: location) { result in
+            switch result {
+            case .failure(let error):
+                // Package index fails to produce result so it cannot be the provider
+                callback(.failure(error), nil)
+            case .success(let metadata):
+                let package = metadata.package
+                let basicMetadata = PackageCollectionsModel.PackageBasicMetadata(
+                    summary: package.summary,
+                    keywords: package.keywords,
+                    versions: package.versions.map { version in
+                        PackageCollectionsModel.PackageBasicVersionMetadata(
+                            version: version.version,
+                            title: version.title,
+                            summary: version.summary,
+                            createdAt: version.createdAt
+                        )
+                    },
+                    watchersCount: package.watchersCount,
+                    readmeURL: package.readmeURL,
+                    license: package.license,
+                    authors: package.authors,
+                    languages: package.languages
+                )
+                
+                let url = try? temp_await { callback in self.get(callback: callback) }
```

Comment:
temp_await needed?

Reply:
No, if config (i.e., package index URL) doesn't change during lifetime of `PackageIndex`.

Reply:
amended https://github.com/apple/swift-package-manager/pull/4010/commits/c1d364a022d867b5bfcd9f69006a1fc714001414